mirascope/chat/parsers.py

```diff
@@ -0,0 +1,60 @@
+"""Classes for using parsers with Chat APIs."""
+import json
+from typing import Callable, Generator, Optional, Type, Union
+
+from openai.types.chat.chat_completion_message_tool_call import (
+    ChatCompletionMessageToolCall,
+    Function,
+)
+from pydantic import BaseModel
+
+from ..partial import Partial
+from .tools import OpenAITool
+from .types import OpenAIChatCompletionChunk
+from .utils import convert_tools_list_to_openai_tools
+
+
+class PartialToolParser(BaseModel):
+    """A utility class to parse `OpenAIChatCompletionChunk`s into `PartialModels`."""
+
+    tool_calls: list[ChatCompletionMessageToolCall] = []
+    tools: list[Union[Callable, Type[OpenAITool]]] = []
+
+    def from_stream(
+        self, stream: Generator[OpenAIChatCompletionChunk, None, None]
+    ) -> Generator[Partial, None, None]:
+        """Parses a stream of `OpenAIChatCompletionChunk`s into `PartialTools`s."""
+        openai_tools = convert_tools_list_to_openai_tools(self.tools)
+        prompt: Optional[Type[OpenAITool]] = None
+        for chunk in stream:
+            tcchunklist = chunk.tool_calls
+            if not tcchunklist:
+                continue
+            for tcchunk in tcchunklist:
+                if len(self.tool_calls) <= tcchunk.index:
```

Comment:
i tried implementing this but it became harder to read due to the continue. I would have to return a tuple which contains (PartialTool, prompt) to keep track of prompt as a iterate since prompt gets reset each iteration since prompt doesnt get mutated.

Reply:
update: the inner for loop is unnecessary because chunk.tool_calls length is always 1.